### DISCUSSION TITLE
implies ⇒ and equivalent ⇔ operators

### DISCUSSION THREAD

**Discussion by Lakedaemon** [josdejong/mathjs#2660] [Ideas]
math.js implement the logical operators or, and, not, xor.
Why doesn't it implement "implies" ⇒ and "equivalence" ⇔ ? 

Also, is there a way to extend math.js with operators ? with symbols ? (with corresponding eval and toTex callBacks ) ?

**Comment by FSMaxB** [comment]
For the question about extending mathjs with custom operators, take a look at #467. Currently you would have to modify the parser yourself manually.

The logical operators `or`, `and`, `not` and `xor` work on boolean values. `⇒` and `⇔` are usually used to express a relation between expression for writing mathematical proofs, I'm not quite sure how they would work as operators in mathjs.

Because mathjs isn't a computer algebra system, comparing expressions would essentially come down to evaluating it and comparing the value. Wouldn't `⇔` be identical to `==` in this case? And how would `⇒` work? Maybe like this:

| `⇒` | `true` | `false` |
| --- | --- | --- |
| **`true`** | `true` | `false` |
| **`false`** | `true` | `true` |

Not quite sure what you intend those operators to do.

**Comment by Lakedaemon** [comment]
Thanks for the pointer to #467 

I'm trying to implement a (simple) proof checker. 
I'm doing (basic) symbolic computations on trees (tree transformations).
comparing expressions is somewhat akin (but not quite) to evaluating it and comparing the value

⇔ is quite identical to == but doesn't expand to TeX with the same symbol and only operates on logical propositions/boolean values
⇒ is implemented as you wrote

At the moment, I use them through math.parse and FunctionNode
It works but the input is quite human unfriendly especially with deeply nested stuff like 
imply(A, imply(B, C))  instead of A ⇒ (B  ⇒ C) etc...

**Comment by FSMaxB** [comment]
Implementing these operators seems quite feasible. But ultimately it's @josdejong's decision.

If Jos accepts this the only issue that remains in my opinion, is the choice of the operators. I don't think it is a good idea to introduce unicode operators into mathjs. So maybe `<=>` and `=>` can be used, I'm not quite sure about that though.

**Comment by josdejong** [comment]
Interesting discussion, thanks.

Some ideas that crossed my mind:
- Nicest would be to make the parser customizable, allow adding new operators. That would give you the freedom to introduce new operators like `<=>` and `=>` yourself. I think there is already an open issue for this.
- I agree with Max that it's probably best to create these operators using regular ascii symbols instead of unicode ⇒ and ⇔. But that could depend on your use case of course.
- Until we have support for symbolic computation, I don't see any use for implementing these operators in math.js: you can't do anything with it. Support for symbolic computation has been worked on already but is on the back burner for now.

**Comment by Lakedaemon** [comment]
Point 1 would be awesome. But I can do without points 1, 2, 3 for the moment (thanks to FunctioNode)
I'll keep experimenting and get back to you if I manage to produce something interesting with math.js and basic symbolic computations
